src/main.rs

```diff
@@ -69,29 +77,73 @@ impl Client {
     // Method should return an error when a connection already exists.
     // The client should send a handshake to the server.
     fn open(&mut self, addr: &str, server: Server) -> CommsResult<()> {
-        todo!()
+        if let std::collections::hash_map::Entry::Vacant(e) =
+            self.connections.entry(String::from(addr))
+        {
+            e.insert(Connection::Open(server));
+            let res = self.send(addr, Message::new(MessageType::Handshake, self.ip.clone()));
+            match res {
+                Err(x) => Err(x),
+                _ => Ok(()),
+            }
+        } else {
+            Err(CommsError::ConnectionExists(String::from(addr)))
+        }
     }
 
     // Sends the provided message to the server at the given `addr`.
     // Can only send messages through open connections. If the server
     // responds with a ServerLimitReached error, its corresponding connection
     // should be closed.
     fn send(&mut self, addr: &str, msg: Message) -> CommsResult<Response> {
-        // server.receive(msg)
-        todo!()
+        #[warn(clippy::map_entry)]
+        if !self.connections.contains_key(&String::from(addr)) {
+            return Err(CommsError::ConnectionNotFound(String::from(addr)));
+        }
+
+        let con = &self.connections[&String::from(addr)];
+        match con {
+            Connection::Open(serv) => {
+                let mut server = serv.clone();
+                let res = server.receive(msg);
+                match res {
+                    Result::Err(CommsError::ServerLimitReached(strn)) => {
+                        *self.connections.get_mut(&String::from(addr)).unwrap() =
+                            Connection::Closed;
+                        Result::Err(CommsError::ServerLimitReached(strn))
+                    }
+                    _ => {
+                        *self.connections.get_mut(&String::from(addr)).unwrap() =
+                            Connection::Open(server);
+                        res
+                    }
+                }
+            }
+            Connection::Closed => Err(CommsError::ConnectionClosed(String::from(addr))),
+        }
     }
 
     // Returns whether the connection to `addr` exists and has
     // the `Open` status.
     #[allow(dead_code)]
     fn is_open(&self, addr: &str) -> bool {
-        todo!()
+        let con = &self.connections[&String::from(addr)];
```

Comment:
Operator `[]` panikuje, gdy nie ma danego klucza w mapie. To dosyć nieoczekiwane działanie w tej metodzie.